content/en/blog/2022-06-27-unikraft-synchronization.md

```diff
@@ -0,0 +1,298 @@
+A simple implementation can be done by using the spinlock.
+But to get better performance we have implemented it using the atomic operation ```__atomic_compare_exchange_n```.
+
+Before implementing the reader-writer lock we need to consider the following problems:
+* Starvation of the writers due to readers.
+* Lost Wake-up problem.
+
+To solve the first problem, we give the higher priority to the writers i.e.  in the lock function when the reader tries to acquire the lock it first checks if there are any pending writers, if yes it sleeps in the queue.
+Also in unlock function for both reader and writer lock we first check if there are any pending writers if yes then we simply wake up the writers instead of the readers.
+To solve the Second problem we can use the sleep with the timeout.
+
+The reader-writer lock is implemented using six basic functions which are essentially ```uk_rwlock_rlock, uk_rwlock_runlock, uk_rwlock_wlock, uk_rwlock_wunlock, uk_rwlock_upgrade, uk_rwlock_downgrade```.
+To implement these functions we need three flags which are
+* UK_RWLOCK_READ : 1 if the lock is held by readers, 0 if the lock is held by writers
+* UK_RWLOCK_READ_WAITERS : 1 if readers are waiting for the lock, otherwise 0
+* UK_RWLOCK_WRITE_WAITERS : 1 if writers are waiting for the lock, otherwise 0
+
+These flags form the lower three bits of the lock variable.
+The remaining bits have different meaning based on UK_RWLOCK_READ flag.
+When this flag is 1 i.e. lock is held by reader then upper bits store the count of the readers holding this lock.
+Otherwise when the lock is held by the writer these bits store the owner of the lock.
+Now instead of using the struct thread pointer, we distinguish the owners of the lock by using the address of the bottom of the stack.
+Since each thread has a unique stack which aligned to page size we get lower 12 bits to store the flags.
+Along with this, there are separate queues for pending readers and pending writers.
+
+Now we describe each method
+
+### Reader Lock
+The reader checks if it can acquire the lock, i.e. there are no current or pending writers.
+if yes then it increments the readers' count, otherwise it sets the ```UK_RWLOCK_READ_WAITERS``` flag and sleeps in the shared queue.
+
+```C
+uk_rwlock_rlock(uk_rwlock *rwl)
+{
+	for(;;) {
+		v = rwl->rwlock;
+		if(_rw_can_read(v)) {
+			setv = v + UK_RW_ONE_READER;
+			if(ukarch_compare_exchange_sync(&rwl->rwlock, v, setv) == setv) {
+				break;
+			}
+			continue;
+		}
+
+		if(!(rwl->rwlock & UK_RWLOCK_READ_WAITERS)) {
+			ukarch_or(&rwl->rwlock, UK_RWLOCK_READ_WAITERS);
+		}
+
+		uk_waitq_wait_event(&rwl->shared, _rw_can_read(rwl->rwlock));
+	}
+}
+```
+
+### Writer Lock
+The writer checks if it can write the lock, i.e. the lock is not owned by the reader or the writer.
+If yes then it unsets the ```UK_RWLOCK_READ``` flag, otherwise it sets the ```UK_RWLOCK_WRITE_WAITERS``` flag and sleeps in the exclusivequeue.
```

Comment:
```suggestion
If it can, then it unsets the `UK_RWLOCK_READ` flag, otherwise it sets the `UK_RWLOCK_WRITE_WAITERS` flag and waits in the exclusive queue.
```